lib/util/fonts/snapToAvailableFontProperties.js

```diff
@@ -0,0 +1,127 @@
+// This implementation follows the font matching algorithm from https://www.w3.org/TR/css-fonts-3/#font-style-matching
+var _ = require('lodash');
+var resolveFontWeight = require('./resolveFontWeight');
+
+var fontStretchValues = [
+    'ultra-condensed',
+    'extra-condensed',
+    'condensed',
+    'semi-condensed',
+    'normal',
+    'semi-expanded',
+    'expanded',
+    'extra-expanded',
+    'ultra-expanded'
+];
+
+// font-style lookup order
+var styleLookupOrder = {
+    normal: ['normal', 'oblique', 'italic'],
+    italic: ['italic', 'oblique', 'normal'],
+    oblique: ['oblique', 'italic', 'normal']
+};
+
+var INITIAL_VALUES = {
+    'font-family': undefined,
+    'font-stretch': 'normal',
+    'font-weight': 400,
+    'font-style': 'normal'
+};
+
+function ascendng(a, b) {
+    return a - b;
+}
+
+function snapToAvailableFontProperties(fontFaceDeclarations, propsToSnap) {
+    if (!Array.isArray(fontFaceDeclarations)) {
+        throw new TypeError('fontFaceDeclarations must be an array');
+    }
+
+    if (typeof propsToSnap !== 'object' || Array.isArray(propsToSnap) || propsToSnap === null) {
+        throw new TypeError('propsToSnap must be an object');
+    }
+
+
+    // Fill in initial values for missing properties
+    fontFaceDeclarations = fontFaceDeclarations.map(function (fontFaceDeclaration) {
+        return _.defaults({}, fontFaceDeclaration, INITIAL_VALUES);
+    });
+    propsToSnap = _.defaults({}, propsToSnap, INITIAL_VALUES);
+
+    // System font, we can't know about the full properties. Early exit
+    if (typeof propsToSnap['font-family'] === 'undefined') {
+        return propsToSnap;
+    }
+
+    // Match font-family first
+    var familymatches = _.filter(fontFaceDeclarations, { 'font-family': propsToSnap['font-family'] });
+
+    // No match for font-family. Probably not a web font. Early exit
+    if (familymatches.length === 0) {
+        return propsToSnap;
+    }
+
+    // Find the best font-stretch
+    var stretchStartIndex = fontStretchValues.indexOf(propsToSnap['font-stretch']);
+    var stretchGroups = _.groupBy(familymatches, 'font-stretch');
+
+    var firstHalf, lastHalf, stretchSearchOrder;
+    var stretchMatches = [];
+
+    if (stretchStartIndex <= fontStretchValues.indexOf('normal')) {
+        // When value is 'normal' or lower, check denser values first, then less dense
+        firstHalf = fontStretchValues.slice(0, stretchStartIndex + 1);
+        lastHalf  = fontStretchValues.slice((stretchStartIndex + 1) - fontStretchValues.length);
```

Comment:
I'm using `stretchStartIndex + 1` in multiple places as a single construct, so I opted to keep these to make it easier to recognise